b2sdk/_test_manager/api.py

```diff
@@ -0,0 +1,188 @@
+            bucket = self.get_bucket_by_name(bucket)
+
+        files_leftover = False
+        file_versions = bucket.ls(latest_only=False, recursive=True)
+
+        for file_version_info, _ in file_versions:
+            if file_version_info.file_retention:
+                if file_version_info.file_retention.mode == RetentionMode.GOVERNANCE:
+                    print('Removing retention from file version:', file_version_info.id_)
+                    self.update_file_retention(
+                        file_version_info.id_, file_version_info.file_name,
+                        NO_RETENTION_FILE_SETTING, True
+                    )
+                elif file_version_info.file_retention.mode == RetentionMode.COMPLIANCE:
+                    if file_version_info.file_retention.retain_until > current_time_millis():  # yapf: disable
+                        print(
+                            'File version: %s cannot be removed due to compliance mode retention' %
+                            (file_version_info.id_,)
+                        )
+                        files_leftover = True
+                        continue
+                elif file_version_info.file_retention.mode == RetentionMode.NONE:
+                    pass
+                else:
+                    raise ValueError(
+                        'Unknown retention mode: %s' % (file_version_info.file_retention.mode,)
+                    )
+            if file_version_info.legal_hold.is_on():
+                print('Removing legal hold from file version:', file_version_info.id_)
+                self.update_file_legal_hold(
+                    file_version_info.id_, file_version_info.file_name, LegalHold.OFF
+                )
+            print('Removing file version:', file_version_info.id_)
+            try:
+                self.delete_file_version(file_version_info.id_, file_version_info.file_name)
+            except FileNotPresent:
+                print(
+                    'It seems that file version %s has already been removed' %
+                    (file_version_info.id_,)
+                )
+
+        if files_leftover:
+            print('Unable to remove bucket because some retained files remain')
+        else:
+            print('Removing bucket:', bucket.name)
+            try:
+                self.delete_bucket(bucket)
+            except BucketIdNotFound:
+                print('It seems that bucket %s has already been removed' % (bucket.name,))
+        print()
+
+    def clean_buckets(self) -> None:
+        for bucket in self.buckets:
+            with contextlib.suppress(BucketIdNotFound, v3BucketIdNotFound, NonExistentBucket):
+                self.clean_bucket(bucket)
+        self.buckets = []
+
+    def clean_all_buckets(self) -> None:
+        buckets = self.list_buckets()
+        print(f'Total bucket count: {len(buckets)}')
+
+        for bucket in buckets:
+            if not bucket.name.startswith(BUCKET_NAME_PREFIX):
```

Comment:
Do we really want an operation like that? What about other runners running in parallel on the same set of keys?

Reply:
As I understand it, this function is intended as a kind of garbage collector after all the tests have been performed in order to collect the "leftovers". You are right about if several commits run the tests there may be unwanted deletions. It seems to me that it is enough here to add the first 10-15 characters from the SHA of the commit to our `BUCKET_NAME_PREFIX`. This way we will be sure that after all finished runs if there are any buckets left the garbage collector will remove them. What do you think?